aws/rust-runtime/aws-http/src/content_encoding.rs

```diff
@@ -0,0 +1,414 @@
+        for (header_name, header_value) in trailer_map.into_iter() {
+            match header_name {
+                // New name, new value
+                Some(header_name) => {
+                    if already_wrote_first_trailer {
+                        // First trailer shouldn't have a preceding CRLF, but every trailer after it should
+                        trailers.write_str(CRLF).unwrap();
+                    } else {
+                        already_wrote_first_trailer = true;
+                    }
+
+                    trailers.write_str(header_name.as_str()).unwrap();
+                    trailers.write_char(':').unwrap();
+                }
+                // Same name, new value
+                None => {
+                    trailers.write_char(',').unwrap();
+                }
+            }
+            trailers.write_str(header_value.to_str().unwrap()).unwrap();
+        }
+    }
+
+    // Write CRLF to end the body
+    trailers.write_str(CRLF).unwrap();
+    // If we wrote at least one trailer, we need to write an extra CRLF
+    if total_length_of_trailers_in_bytes != 0 {
+        trailers.write_str(CRLF).unwrap();
+    }
+
+    trailers.into()
+}
+
+impl Body for AwsChunkedBody<Inner> {
+    type Data = Bytes;
+    type Error = aws_smithy_http::body::Error;
+
+    fn poll_data(
+        self: Pin<&mut Self>,
+        cx: &mut Context<'_>,
+    ) -> Poll<Option<Result<Self::Data, Self::Error>>> {
+        tracing::info!("polling AwsChunkedBody");
+        let mut this = self.project();
+
+        match *this.state {
+            AwsChunkedBodyState::WritingChunkSize => match this.inner.poll_data(cx) {
+                Poll::Ready(Some(Ok(data))) => {
+                    // A chunk must be prefixed by chunk size in hexadecimal
+                    tracing::info!("writing chunk size and start of chunk");
+                    *this.state = AwsChunkedBodyState::WritingChunk;
+                    let total_chunk_size = this
+                        .options
+                        .chunk_length
+                        .or(this.options.stream_length)
+                        .unwrap_or_default();
+                    Poll::Ready(Some(Ok(prefix_with_chunk_size(data, total_chunk_size))))
+                }
+                Poll::Ready(None) => {
+                    tracing::info!("chunk was empty, writing last-chunk");
+                    *this.state = AwsChunkedBodyState::WritingTrailers;
+                    Poll::Ready(Some(Ok(Bytes::from("0\r\n"))))
+                }
+                Poll::Ready(Some(Err(e))) => Poll::Ready(Some(Err(e))),
+                Poll::Pending => Poll::Pending,
+            },
+            AwsChunkedBodyState::WritingChunk => match this.inner.poll_data(cx) {
+                Poll::Ready(Some(Ok(mut data))) => {
+                    tracing::info!("writing rest of chunk data");
+                    Poll::Ready(Some(Ok(data.copy_to_bytes(data.len()))))
+                }
```

Comment:
`data` _is_ bytes I think? so I don't think we need this copy?

Reply:
we don't, fixed it